tracing/tracing.go

```diff
@@ -0,0 +1,123 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package tracing
+
+import (
+	"context"
+
+	srvconfig "github.com/containerd/containerd/services/server/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+)
+
+// InitOpenTelemetry read config and initializes otel middleware, sets the exporter
+// propagator and global tracer provider.
+func InitOpenTelemetry(config *srvconfig.Config) (func(), error) {
+	ctx := context.Background()
+
+	// Check if tracing is configured
+	if config.OpenTelemetry == *new(srvconfig.OpenTelemetryConfig) {
+		logrus.Info("OpenTelemetry configuration is not found in config, tracing will be disabled")
+		return nil, nil
+	}
+
+	// Validate configuration
+	err := config.OpenTelemetry.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid open telemetry configuration")
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			// Service name used to displace traces in backends
+			semconv.ServiceNameKey.String("containerd"),
+		),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create resource")
+	}
+
+	// Configure OTLP trace exporter and set it up to connect to OpenTelemetry collector
+	// running on a local host.
+	ctrdTraceExporter, err := otlptracegrpc.New(ctx,
+		otlptracegrpc.WithInsecure(),
+		otlptracegrpc.WithEndpoint(config.OpenTelemetry.ExporterEndPoint),
+		otlptracegrpc.WithDialOption(grpc.WithBlock()),
```

Comment:
Is it fine to use a remote location as its endpoint? If so, is it appropriate to force non-encrypted transport?

Reply:
Probably should be config coming from toml file insteading enforcing it here. And yes, remote connection should be allowed. But the combination of endpoint and whether to connect it in insecure mode should be coming from config i guess.

Reply:
For now will remove the inSecure option.